Product/Production/Common/CONNECTCoreLib/src/main/java/gov/hhs/fha/nhinc/callback/opensaml/CertificateUtil.java

```diff
@@ -76,4 +98,137 @@ private void closeStream(Closeable stream) {
+            byte[] akiBytes = cert.getExtensionValue(AUTHORITY_KEY_ID);
+            try {
+                if (akiBytes != null) {
+                    DERDecoder extValA = new DERDecoder(akiBytes);
+                    extValA.skip(6);
+                    aki = Hex.encodeHexString(extValA.getBytes(20));
+                }
+            } catch (WSSecurityException e) {
+                LOG.error("Unable to decode certificate authority {}", e.getLocalizedMessage(), e);
+            }
+        }
+        return aki;
+    }
+
+    public static String getCertKeyIdSubject(Certificate cert) {
+        return getCertKeyIdSubject((X509Certificate) cert);
+    }
+    public static String getCertKeyIdSubject(X509Certificate cert) {
+        String ski = null;
+        if (null != cert) {
+            byte[] subjectKeyID = cert.getExtensionValue(SUBJECT_KEY_ID);
+            try {
+                if (subjectKeyID != null) {
+                    // this logic extracts from CryptoBase class inside wss4j
+                    DERDecoder extVal = new DERDecoder(subjectKeyID);
+                    extVal.expect(DERDecoder.TYPE_OCTET_STRING); // ExtensionValue OCTET STRING
+                    extVal.getLength(); // leave this method alone. getlength modify array position.
+                    extVal.expect(DERDecoder.TYPE_OCTET_STRING); // KeyIdentifier OCTET STRING
+                    int keyIDLen = extVal.getLength();
+                    byte[] hexValue = extVal.getBytes(keyIDLen);
+                    ski = Hex.encodeHexString(hexValue);
+                }
+            } catch (WSSecurityException e) {
+                LOG.error("Unable to decode certificate subject: {}", e.getLocalizedMessage(), e);
+            }
+        }
+        return ski;
+    }
+
+    public static String getCertSubjectCN(Certificate cert) {
+        return getCertSubjectCN((X509Certificate) cert);
+    }
+    public static String getCertSubjectCN(X509Certificate cert) {
+        if (null != cert) {
+            try {
+                LdapName ldapDN = new LdapName(cert.getSubjectX500Principal().getName());
+                for (Rdn rdn : ldapDN.getRdns()) {
+                    if (rdn.getType().equals(COMMON_NAME)) {
+                        return (String) rdn.getValue();
+                    }
+                }
+            } catch (InvalidNameException ex) {
+                LOG.error("{}", ex.getMessage(), ex);
+            }
+        }
+        return null;
+    }
+
+    public static boolean isInChain(Certificate cert, List<Certificate> chain) {
+        String lookSKI = getCertKeyIdSubject(cert);
+        if (null != lookSKI) {
+            for (Certificate link : chain) {
+                String foundSKI = getCertKeyIdSubject(link);
+                if (lookSKI.equals(foundSKI)) {
+                    return true;
+                }
+            }
+        }
+        return false;
+    }
+
+    public static long getDaysOfExpiration(Date dateExpire) {
```

Comment:
Can we rename this method to say "getExpirationTimeInMillis"

Reply:
no this is returning the in days; i can rename "getExpirationTimeInDays"